### PR TITLE
Support `VersionNumber`

### LINKED ISSUE #200: VersionNumbers don't serialize
and give a confusing error:
```julia
julia> using Arrow

julia> table = [(; version = v"1")]
1-element Vector{NamedTuple{(:version,), Tuple{VersionNumber}}}:
 (version = v"1.0.0",)

julia> Arrow.tobuffer(table)
ERROR: UndefVarError: T not defined
Stacktrace:
  [1] Arrow.ArrowTypes.ArrowKind(#unused#::Type{Tuple{}})
    @ Arrow.ArrowTypes ~/.julia/packages/Arrow/ggVa0/src/ArrowTypes/src/ArrowTypes.jl:225
  [2] arrowvector(::Type{Tuple{}}, x::Arrow.ArrowTypes.ToArrow{Tuple{}, Arrow.ToStruct{Tuple{Vararg{Union{UInt64, String}, N} where N}, 4, Vector{VersionNumber}}}, i::Int64, nl::Int64, fi::Int64, de::Dict{Int64, Any}, ded::Vector{Arrow.DictEncoding}, meta::Dict{String, String}; kw::Base.Iterators.Pairs{Symbol, Union{Nothing, Integer}, NTuple{6, Symbol}, NamedTuple{(:dictencode, :maxdepth, :compression, :largelists, :denseunions, :dictencodenested), Tuple{Bool, Int64, Nothing, Bool, Bool, Bool}}})
    @ Arrow ~/.julia/packages/Arrow/ggVa0/src/arraytypes/arraytypes.jl:85
  [3] arrowvector(x::Arrow.ToStruct{Tuple{Vararg{Union{UInt64, String}, N} where N}, 4, Vector{VersionNumber}}, i::Int64, nl::Int64, fi::Int64, de::Dict{Int64, Any}, ded::Vector{Arrow.DictEncoding}, meta::Nothing; dictencoding::Bool, dictencode::Bool, maxdepth::Int64, kw::Base.Iterators.Pairs{Symbol, Union{Nothing, Bool}, NTuple{4, Symbol}, NamedTuple{(:compression, :largelists, :denseunions, :dictencodenested), Tuple{Nothing, Bool, Bool, Bool}}})
    @ Arrow ~/.julia/packages/Arrow/ggVa0/src/arraytypes/arraytypes.jl:69
  [4] (::Arrow.var"#39#40"{Base.Iterators.Pairs{Symbol, Union{Nothing, Integer}, NTuple{6, Symbol}, NamedTuple{(:dictencode, :maxdepth, :compression, :largelists, :denseunions, :dictencodenested), Tuple{Bool, Int64, Nothing, Bool, Bool, Bool}}}, Vector{VersionNumber}, Int64, Int64, Dict{Int64, Any}, Vector{Arrow.DictEncoding}})(j::Int64)
    @ Arrow ./none:0
  [5] iterate
    @ ./generator.jl:47 [inlined]
  [6] collect_to!(dest::Vector{Arrow.Primitive{UInt32, A} where A}, itr::Base.Generator{UnitRange{Int64}, Arrow.var"#39#40"{Base.Iterators.Pairs{Symbol, Union{Nothing, Integer}, NTuple{6, Symbol}, NamedTuple{(:dictencode, :maxdepth, :compression, :largelists, :denseunions, :dictencodenested), Tuple{Bool, Int64, Nothing, Bool, Bool, Bool}}}, Vector{VersionNumber}, Int64, Int64, Dict{Int64, Any}, Vector{Arrow.DictEncoding}}}, offs::Int64, st::Int64)
    @ Base ./array.jl:724
  [7] collect_to!(dest::Vector{Arrow.Primitive{UInt32, Arrow.ToStruct{UInt32, 1, Vector{VersionNumber}}}}, itr::Base.Generator{UnitRange{Int64}, Arrow.var"#39#40"{Base.Iterators.Pairs{Symbol, Union{Nothing, Integer}, NTuple{6, Symbol}, NamedTuple{(:dictencode, :maxdepth, :compression, :largelists, :denseunions, :dictencodenested), Tuple{Bool, Int64, Nothing, Bool, Bool, Bool}}}, Vector{VersionNumber}, Int64, Int64, Dict{Int64, Any}, Vector{Arrow.DictEncoding}}}, offs::Int64, st::Int64)
    @ Base ./array.jl:732
  [8] collect_to_with_first!(dest::Vector{Arrow.Primitive{UInt32, Arrow.ToStruct{UInt32, 1, Vector{VersionNumber}}}}, v1::Arrow.Primitive{UInt32, Arrow.ToStruct{UInt32, 1, Vector{VersionNumber}}}, itr::Base.Generator{UnitRange{Int64}, Arrow.var"#39#40"{Base.Iterators.Pairs{Symbol, Union{Nothing, Integer}, NTuple{6, Symbol}, NamedTuple{(:dictencode, :maxdepth, :compression, :largelists, :denseunions, :dictencodenested), Tuple{Bool, Int64, Nothing, Bool, Bool, Bool}}}, Vector{VersionNumber}, Int64, Int64, Dict{Int64, Any}, Vector{Arrow.DictEncoding}}}, st::Int64)
    @ Base ./array.jl:702
  [9] collect(itr::Base.Generator{UnitRange{Int64}, Arrow.var"#39#40"{Base.Iterators.Pairs{Symbol, Union{Nothing, Integer}, NTuple{6, Symbol}, NamedTuple{(:dictencode, :maxdepth, :compression, :largelists, :denseunions, :dictencodenested), Tuple{Bool, Int64, Nothing, Bool, Bool, Bool}}}, Vector{VersionNumber}, Int64, Int64, Dict{Int64, Any}, Vector{Arrow.DictEncoding}}})
    @ Base ./array.jl:683
 [10] _totuple
    @ ./tuple.jl:331 [inlined]
 [11] Tuple
    @ ./tuple.jl:303 [inlined]
 [12] arrowvector(::Arrow.ArrowTypes.StructKind, x::Vector{VersionNumber}, i::Int64, nl::Int64, fi::Int64, de::Dict{Int64, Any}, ded::Vector{Arrow.DictEncoding}, meta::Nothing; kw::Base.Iterators.Pairs{Symbol, Union{Nothing, Integer}, NTuple{6, Symbol}, NamedTuple{(:dictencode, :maxdepth, :compression, :largelists, :denseunions, :dictencodenested), Tuple{Bool, Int64, Nothing, Bool, Bool, Bool}}})
    @ Arrow ~/.julia/packages/Arrow/ggVa0/src/arraytypes/struct.jl:93
 [13] arrowvector(::Type{VersionNumber}, x::Vector{VersionNumber}, i::Int64, nl::Int64, fi::Int64, de::Dict{Int64, Any}, ded::Vector{Arrow.DictEncoding}, meta::Nothing; kw::Base.Iterators.Pairs{Symbol, Union{Nothing, Integer}, NTuple{6, Symbol}, NamedTuple{(:dictencode, :maxdepth, :compression, :largelists, :denseunions, :dictencodenested), Tuple{Bool, Int64, Nothing, Bool, Bool, Bool}}})
    @ Arrow ~/.julia/packages/Arrow/ggVa0/src/arraytypes/arraytypes.jl:85
 [14] arrowvector(x::Vector{VersionNumber}, i::Int64, nl::Int64, fi::Int64, de::Dict{Int64, Any}, ded::Vector{Arrow.DictEncoding}, meta::Nothing; dictencoding::Bool, dictencode::Bool, maxdepth::Int64, kw::Base.Iterators.Pairs{Symbol, Union{Nothing, Bool}, NTuple{4, Symbol}, NamedTuple{(:compression, :largelists, :denseunions, :dictencodenested), Tuple{Nothing, Bool, Bool, Bool}}})
    @ Arrow ~/.julia/packages/Arrow/ggVa0/src/arraytypes/arraytypes.jl:69
 [15] toarrowvector(x::Vector{VersionNumber}, i::Int64, de::Dict{Int64, Any}, ded::Vector{Arrow.DictEncoding}, meta::Nothing; compression::Nothing, kw::Base.Iterators.Pairs{Symbol, Integer, NTuple{5, Symbol}, NamedTuple{(:largelists, :denseunions, :dictencode, :dictencodenested, :maxdepth), Tuple{Bool, Bool, Bool, Bool, Int64}}})
    @ Arrow ~/.julia/packages/Arrow/ggVa0/src/arraytypes/arraytypes.jl:36
 [16] (::Arrow.var"#107#108"{Dict{Int64, Any}, Bool, Nothing, Bool, Bool, Bool, Int64, Vector{Arrow.DictEncoding}, Vector{Type}, Vector{Any}})(col::Vector{VersionNumber}, i::Int64, nm::Symbol)
    @ Arrow ~/.julia/packages/Arrow/ggVa0/src/write.jl:243
 [17] eachcolumn
    @ ~/.julia/packages/Tables/YzCZp/src/utils.jl:70 [inlined]
 [18] toarrowtable(x::Vector{NamedTuple{(:version,), Tuple{VersionNumber}}}, dictencodings::Dict{Int64, Any}, largelists::Bool, compress::Nothing, denseunions::Bool, dictencode::Bool, dictencodenested::Bool, maxdepth::Int64)
    @ Arrow ~/.julia/packages/Arrow/ggVa0/src/write.jl:242
 [19] macro expansion
    @ ~/.julia/packages/Arrow/ggVa0/src/write.jl:136 [inlined]
 [20] macro expansion
    @ ./task.jl:387 [inlined]
 [21] write(io::IOBuffer, source::Vector{NamedTuple{(:version,), Tuple{VersionNumber}}}, writetofile::Bool, largelists::Bool, compress::Nothing, denseunions::Bool, dictencode::Bool, dictencodenested::Bool, alignment::Int64, maxdepth::Int64, ntasks::Float64)
    @ Arrow ~/.julia/packages/Arrow/ggVa0/src/write.jl:129
 [22] #write#96
    @ ~/.julia/packages/Arrow/ggVa0/src/write.jl:92 [inlined]
 [23] write
    @ ~/.julia/packages/Arrow/ggVa0/src/write.jl:92 [inlined]
 [24] #tobuffer#3
    @ ~/.julia/packages/Arrow/ggVa0/src/utils.jl:206 [inlined]
 [25] tobuffer(data::Vector{NamedTuple{(:version,), Tuple{VersionNumber}}})
    @ Arrow ~/.julia/packages/Arrow/ggVa0/src/utils.jl:205
 [26] top-level scope
    @ REPL[5]:1
```

I tried various versions of
```julia
using Arrow: ArrowTypes

ArrowTypes.ArrowKind(::Type{VersionNumber}) = ArrowTypes.ListKind()
ArrowTypes.JuliaType(::Val{:VersionNumber}) = VersionNumber
ArrowTypes.arrowname(::Type{:VersionNumber}) = :VersionNumber
ArrowTypes.toarrow(v::VersionNumber) = string(v)
ArrowTypes.fromarrow(::Type{VersionNumber}, v::String) = VersionNumber(v)
```
but did not have success.

### PR DESCRIPTION
**Pull request by omus**
Fixes #200 by adding round-trip support for `VersionNumber`. Requires #201.

### CODE REVIEW

**Comment by omus** [comment]
CI expected to fail due to #202

**Comment by omus** [comment]
CI is disabled for branches not based off of `main`. Will rebase once `cv/empty-tuple` is merged.

**Review by jrevels — APPROVED** [review]
LGTM pending clean rebase on top of main post-merge of #201 

We should tag a release after this is merged IMO.

**Comment by jrevels** [comment]
now that #201 is merged looks like GitHub auto switched this PR's base branch to `main` - will closing/reopening trigger CI? let's see

File: test/runtests.jl
Comment (jrevels) [inline test/runtests.jl:350]:
```suggestion
    @test eltype(tbl.col1) == VersionNumber
```

**Pull request merged**
